### PR TITLE
HPCC-24237 LdapSecurity getting and setting permissions for Azure AD

### PR DESCRIPTION
**Pull request by mayx**
- Add server control to search function to be able to retrieve only
  the Discretionary ACL of nTSecurityDescriptor
- Add server control to modify function to be able to modify
  only the discretionary part of nTSecurityDescriptor

Signed-off-by: Yanrui Ma <yanrui.ma@lexisnexisrisk.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [x] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [x] Scalability
  - [x] Performance
  - [x] Security
  - [x] Thread-safety
  - [x] Cloud-compatibility
  - [x] Premature optimization
  - [x] Existing deployed queries will not be broken
  - [x] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [x] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Comment by mayx** [comment]
@RussWhitehead @afishbeck please review...

File: system/security/LdapSecurity/ldapconnection.cpp [hunk 1]
```diff
@@ -4688,7 +4702,33 @@ class CLdapClient : implements ILdapClient, public CInterface
...
 
 private:
+    class SDServerCtlWrapper
+    {
+    public:
+        LDAPControl **ctls = nullptr;
+        LDAPControl* ctl = nullptr;
+        StringBuffer oidbuf, valbuf;
 
+        SDServerCtlWrapper(bool isAzureAD)
```
Comment (afishbeck) [inline system/security/LdapSecurity/ldapconnection.cpp:4712]: isAzureAD never actually seems to be checked.  Is this backward compatible?

File: system/security/LdapSecurity/ldapconnection.cpp [hunk 2]
```diff
@@ -4688,7 +4702,33 @@ class CLdapClient : implements ILdapClient, public CInterface
...
+        {
+            oidbuf.append("1.2.840.113556.1.4.801");
```
Comment (afishbeck) [inline system/security/LdapSecurity/ldapconnection.cpp:4714]: Maybe a brief comment about what these numbers represent?

**Review by afishbeck — COMMENTED** [review]
@mayx a couple questions.

File: system/security/LdapSecurity/ldapconnection.cpp (hunk 1 above)
Comment (mayx) [reply system/security/LdapSecurity/ldapconnection.cpp:4712]: Meant to check it, guess forgot. Will add...

**Comment by mayx** [comment]
@afishbeck change made per suggestion, please review again...

**Review by afishbeck — APPROVED** [review]
@mayx looks good

**Comment by RussWhitehead** [comment]
@mayx Looks good

**Comment by mayx** [comment]
@afishbeck made a small change per discussion, please review again...

**Review by afishbeck — APPROVED** [review]
@mayx looks good.

**Comment by afishbeck** [comment]
@richardkchapman Some minor improvements.  Azure active directory is now based on configuration and the Azure OID is added as a define to make it clearer.

**Pull request merged**